File: app/schemas/video.py
"""Pydantic schemas for video upload API endpoints."""

from typing import Optional

from pydantic import BaseModel, Field


class PresignedUrlRequest(BaseModel):
    """Request schema for generating a presigned upload URL."""

    file_name: str = Field(..., description="Original file name")
    content_type: str = Field(
        ...,
        pattern=r"^video/(mp4|quicktime|x-msvideo)$",
        description="MIME type of the video file",
    )


class PresignedUrlResponse(BaseModel):
    """Response schema for presigned upload URL."""

    upload_url: str = Field(..., description="Presigned S3 upload URL")
    file_key: str = Field(..., description="S3 object key for the uploaded file")
    expires_in: int = Field(..., description="URL expiration time in seconds")


class ResolutionResponse(BaseModel):
    """Resolution sub-schema with width and height."""

    width: int
    height: int


class VideoMetadataResponse(BaseModel):
    """Response schema for video metadata after upload."""

    file_key: str
    file_name: str
    file_size_bytes: int
    duration_seconds: float
    resolution: ResolutionResponse
    frame_rate: float
    codec: str
    format: str
    thumbnail_url: Optional[str] = None


class VideoMetadataWithThumbnailResponse(BaseModel):
    """Response schema for video metadata with thumbnail (Requirement 1.7).

    Returns file_name, duration_seconds, resolution, file_size_bytes, and thumbnail_url.
    """

    file_name: str
    duration_seconds: float
    resolution: ResolutionResponse
    file_size_bytes: int
    thumbnail_url: Optional[str] = None


class VideoValidationResponse(BaseModel):
    """Response schema for video validation result."""

    is_valid: bool
    format_ok: bool
    size_ok: bool
    resolution_ok: bool
    frame_rate_ok: bool
    errors: list[str] = Field(default_factory=list)


class QualityCheckResponse(BaseModel):
    """Response schema for video quality check result."""

    brightness_status: str  # "pass" | "warning"
    framing_status: str  # "pass" | "warning"
    resolution_status: str  # "pass" | "warning"
    frame_rate_stability_status: str  # "pass" | "warning"
    brightness_value: float
    swing_arc_visibility_percent: float
    frame_rate_variation_percent: float
    warnings: list[str] = Field(default_factory=list)

File: frontend/src/types/video.ts
/**
 * Type definitions matching backend Pydantic schemas in app/schemas/video.py.
 */

export interface PresignedUrlRequest {
  file_name: string;
  /** MIME type matching ^video/(mp4|quicktime|x-msvideo)$ */
  content_type: string;
}

export interface PresignedUrlResponse {
  upload_url: string;
  file_key: string;
  expires_in: number;
}

export interface ResolutionResponse {
  width: number;
  height: number;
}

export interface VideoMetadataResponse {
  file_key: string;
  file_name: string;
  file_size_bytes: number;
  duration_seconds: number;
  resolution: ResolutionResponse;
  frame_rate: number;
  codec: string;
  format: string;
  thumbnail_url?: string | null;
}


export type VideoInputPolicySeverity = "ok" | "warning" | "error";

export type VideoInputPolicyReason =
  | "video_too_short"
  | "video_longer_than_recommended"
  | "video_too_long"
  | "metadata_unavailable";

export interface VideoInputPolicyResponse {
  accepted: boolean;
  severity: VideoInputPolicySeverity;
  reason?: VideoInputPolicyReason | null;
  duration_sec?: number | null;
  ideal_duration_sec: number;
  recommended_min_duration_sec: number;
  recommended_max_duration_sec: number;
  max_duration_sec: number;
  message: string;
  recommendation: string;
}

export interface VideoMetadataWithThumbnailResponse {
  file_name: string;
  duration_seconds: number;
  resolution: ResolutionResponse;
  file_size_bytes: number;
  thumbnail_url?: string | null;
  input_validation?: VideoInputPolicyResponse | null;
}

export type QualityStatus = "pass" | "warning";

export interface QualityCheckResponse {
  brightness_status: QualityStatus;
  framing_status: QualityStatus;
  resolution_status: QualityStatus;
  frame_rate_stability_status: QualityStatus;
  brightness_value: number;
  swing_arc_visibility_percent: number;
  frame_rate_variation_percent: number;
  warnings: string[];
}

export interface VideoValidationResponse {
  is_valid: boolean;
  format_ok: boolean;
  size_ok: boolean;
  resolution_ok: boolean;
  frame_rate_ok: boolean;
  errors: string[];
}
